### PR TITLE
Replace generic ProgramErrors with program-specific WalletErrors

### PR DESCRIPTION
**Pull request by dgabriele**
- Replaced ProgramErrors with program-specific WalletErrors to make the cause of errors clearer.
- Added docstrings to WalletErrors, based on the format found in Solana Program Library. For example: https://github.com/solana-labs/solana-program-library/blob/master/token/program/src/error.rs
- Small refactoring in Makefile to make it more DRY.

## How Has This Been Tested?
Tests updated to catch new error codes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix breaking (fix that would cause existing functionality to change)

### CODE REVIEW

File: src/model/wallet.rs [hunk 1]
```diff
@@ -402,7 +402,7 @@ impl Wallet {
     fn add_signers(&mut self, signers_to_add: &Vec<(SlotId<Signer>, Signer)>) -> ProgramResult {
         if !self.signers.can_be_inserted(signers_to_add) {
             msg!("Failed to add signers: at least one of the provided slots is already taken");
-            return Err(ProgramError::InvalidArgument);
+            return Err(WalletError::InvalidSlot.into());
```
Comment (bholzman) [inline src/model/wallet.rs:405]: So yeah, thinking that `InvalidSlot` might be too generic given how many times it occurs. The debug messages make it clear what the problem actually was to a developer, but as there doesn't seem to be a way to retrieve the logs when using the `solana-program-test` framework, checking this error code in a unit test might suffer from the original problem you were trying to fix (of a test getting the error it expects but for the wrong reason).

This one could be maybe something like `SlotAlreadyOccupied`, and the next one could be more like `UnknownSigner`, etc.

Comment (dgabriele) [reply src/model/wallet.rs:405]: Sounds good :)

File: src/model/wallet.rs [hunk 2]
```diff
@@ -259,12 +259,12 @@ impl Wallet {
 
         if self.approvals_required_for_config == 0 {
             msg!("Approvals required for config can't be 0");
-            return Err(ProgramError::InvalidArgument);
+            return Err(WalletError::InvalidApproverCount.into());
         }
 
         if self.approval_timeout_for_config.as_secs() == 0 {
```
Comment (bflood-strike) [inline src/model/wallet.rs:265]: shouldn't this be checking it's within the range 1 minute -> max

Comment (bflood-strike) [reply src/model/wallet.rs:265]: I'm talking about this line (265) that is checking the approval_timeout_for config 
` if self.approval_timeout_for_config.as_secs() == 0 {`

File: src/error.rs [hunk 3]
```diff
@@ -3,38 +3,84 @@ use thiserror::Error;
...
+    /// Only one policy config change can be initiated at a time.
     ConcurrentOperationsNotAllowed,
+    /// Simulation of MultisigOp finalization completed normally.
     #[error("Simulation Finished Successfully")]
     SimulationFinished,
     #[error("Whitelist Is Disabled")]
     WhitelistDisabled,
+    /// Cannot whitelist an address when Whitelisting is not enabled.
+    #[error("Whitelisting Status is Off")]
+    WhitelistingStatusOff,
```
Comment (bholzman) [inline src/error.rs:74]: this should be the `WhitelistDisabled` above, not a new error

File: src/model/wallet.rs (hunk 1 above)
Comment (bholzman) [reply src/model/wallet.rs:417]: seems like there are still a lot of `InvalidSlot` errors, I thought you said you were going to change them, particularly where a single instruction is returning the same `InvalidSlot` error for different conditions, which was what you set out to fix here

**Review by bholzman — CHANGES_REQUESTED** [review]

File: src/error.rs (hunk 3 above)
Comment (dgabriele) [reply src/error.rs]: I got it. Fixed everything up now.

**Review by bholzman — CHANGES_REQUESTED** [review]
Just one more change then I think good to go

**Review by bholzman — APPROVED** [review]
🚢

**Review by bholzman — APPROVED** [review]

**Pull request merged**